lib/Conch/Controller/RackLayout.pm

```diff
@@ -173,10 +173,17 @@ sub update ($c) {
         return $c->status(400, { error => 'changing rack_id is not permitted' });
     }
 
-    # cannot alter an occupied layout
-    if (my $device_location = $layout->device_location) {
-        $c->log->debug('Cannot update layout: occupied by device id '.$device_location->device_id);
+    # only permit updating occupied layouts if the hardware_product_id is the only change,
+    # iff it is changing to the device_hardware_product_id.
+    if (my $device = $layout->related_resultset('device_location')
+        ->related_resultset('device')->columns([qw(id hardware_product_id)])->single) {
+      if (($input->{rack_unit_start} and $input->{rack_unit_start} != $layout->rack_unit_start)
+        or ($input->{hardware_product_id}
+          and $input->{hardware_product_id} ne $layout->hardware_product_id
+          and $input->{hardware_product_id} ne $device->hardware_product_id)) {
```

Comment:
Where is the device data populated from again? We pull it from the device reports no?

Reply:
in v3.0 the device is created during POST .../rack:Id_or_name/assignment, with a sku matching the rack_layout it is assigned to. The sku is not changed during device report processing.

If the sku needs to change after the device has been created, you need to do  POST /device/:device_id_or_serial/hardware_product (or .../sku).

So, to change the sku for a located device, change the device sku first, and then the rack layout sku can be changed to match it.

Reply:
Got it. That works for me.